### PR TITLE
TEP-0121: Add status transition details

### PR DESCRIPTION
**Pull request by XinruZhang**
Add an explaination on how a status changes with every retry of a taskRun/customRun. Adding an example here to show the transition when a task fails v/s a task timesout.

Open this PR to continue the discussion in https://github.com/tektoncd/community/pull/882 as Priti will be out next Month

This PR makes some modifications to address https://github.com/tektoncd/community/pull/882#discussion_r1037161026. See the file diff result between the two PRs: https://www.diffchecker.com/tmp7Jmmr

cc @pritidesai @jerop @afrittoli

### CODE REVIEW

File: teps/0121-refine-retries-for-taskruns-and-customruns.md
Comment (pritidesai) [inline teps/0121-refine-retries-for-taskruns-and-customruns.md:159]: @XinruZhang please add this statement back for clarity:

https://github.com/tektoncd/community/pull/882/files#diff-4a8da49e599db9e42576edb3fda3e041fd32e3f502ab42df54f8eff0d6e301b1R160

Comment (XinruZhang) [reply teps/0121-refine-retries-for-taskruns-and-customruns.md:159]: Thanks Priti @pritidesai Added!

**Comment by pritidesai** [comment]
Sounds good to me, thanks @XinruZhang 

Ready to merge as long as the competition time of each attempt in `retriesStatus` remains consistent with the time set by the taskRun controller when it exited with non-zero or timedout. 

/approve

File: teps/0121-refine-retries-for-taskruns-and-customruns.md
Comment (jerop) [inline teps/0121-refine-retries-for-taskruns-and-customruns.md:160]: to be clear, the `CompletionTime` is set in the `RetriesStatus` upon the failure of a retry, but the `CompletionTime` in the `TaskRun` is set only upon the eventual failure of the `TaskRun` (when all retries have been exhausted) -- is that correct?

Comment (XinruZhang) [reply teps/0121-refine-retries-for-taskruns-and-customruns.md]: Yes! Thanks for asking~ rephrased a little bit to make it clearer ❤️

**Review by jerop — APPROVED** [review]

File: teps/0121-refine-retries-for-taskruns-and-customruns.md
Comment (XinruZhang) [reply teps/0121-refine-retries-for-taskruns-and-customruns.md]: Thanks Jerop! Updated.

**Comment by XinruZhang** [comment]
@afrittoli  Hi Andrea, many thanks for your review of #882! Updated the status transition table to explicitly show when the `completionTime` is set. PTAL and let me know if you have more thoughts!

**Comment by XinruZhang** [comment]
@tektoncd/core-maintainers  PTAL, thanks! 🙏🏻

File: teps/0121-refine-retries-for-taskruns-and-customruns.md
Comment (afrittoli) [inline teps/0121-refine-retries-for-taskruns-and-customruns.md:147]: I understand the purpose of this intermediate `reason` but I'm not sure how it will be implemented.
When the `TaskRun` reconciler detects a failure in the underlying `Pod`, it will:

- store the data of the current run in `retriesStatus`
- if there are no retries left, mark the `TaskRun` as failed
- else, clean up the current `TaskRun` status
- set the `reason` to `ToBeRetried`

At this point, the controller could start the next `TaskRun`, but if it does so, the `ToBeRetried` value will never make it into `etcd`. To make the `ToBeRetried` reason visible to watchers we will have to finish the reconcile cycle and start the retry in the next reconcile cycle.

Is this the planned implementation? It seems to me that the reconcile cycle should do ask much as possible to reconcile the resource, rather than stop and wait for the next reconcile. 
I'm guessing that the changes to the `TaskRun` done by the reconcile will trigger another reconcile cycle, but I'm not 100% sure.

**Review by afrittoli — COMMENTED** [review]
/approve

File: teps/0121-refine-retries-for-taskruns-and-customruns.md
Comment (XinruZhang) [reply teps/0121-refine-retries-for-taskruns-and-customruns.md]: Thanks for the catch!

**Comment by jerop** [comment]
all assignees have approved, merging offline

/lgtm

**Pull request merged**
